internal/proxyservice/proxy_service.go

```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
+//	hs.SubserviceConfigs = append(hs.SubserviceConfigs, svc)
+//}
+
+// configsByType holds proxy service configuration in a form that directly
+// corresponds to the ProxyService objects we want to create.  One ProxyService
+// will be created for each entry in http, and one for each entry in tcp.
+type configsByType struct {
+	tcp  []v2.Service
+	http []HTTPServiceConfig
+}
+
+// newConfigsByType converts a slice of v2.Service configs into the configs
+// needed to actually created ProxyServices -- configsByType.  In particular, it
+// takes all the http configs and creates proper HTTPServiceConfig objects out
+// of them -- grouping the raw v2.Service configs by their listenOn property.
+// The remaining services are tcp, and already correspond to the services we'll
+// run.
+// TODO: Eventually it should get moved to the config package, and this
+//   transformation should occur at the border.
+func newConfigsByType(
+	uncheckedConfigs []*v2.Service,
+	availPlugins plugin.AvailablePlugins,
+) configsByType {
+
+	// Get the nil checks out of the way.
+	// TODO: This Config definition itself should be changed to []v2.Service
+	var rawConfigs []v2.Service
+	for _, cfg := range uncheckedConfigs {
+		if cfg == nil {
+			// Hard-coding log here is okay since nils should never occur and
+			// we won't be unit testing this.  Instead, we'll make the change
+			// in the TODO_ above and this code will be deleted then.
+			log.Fatalln("Nil configuration is not allowed!")
+		}
+		rawConfigs = append(rawConfigs, *cfg)
+	}
+
+	// Separate HTTP and TCP configs
+	var httpConfigs []v2.Service
+	var	tcpConfigs []v2.Service
+
+	// TODO: Add proper validation here of the type.  This should moved into
+	//   IsHTTPPlugin, whose API will likely change to returning a type or an
+	//   error
+	for _, cfg := range rawConfigs {
+		if plugin.IsHTTPPlugin(availPlugins, cfg.Connector) {
+			httpConfigs = append(httpConfigs, cfg)
 			continue
 		}
+		tcpConfigs = append(tcpConfigs, cfg)
+	}
+
+	// Group http configs by their ListenOn property
+	httpByListenOn := map[string][]v2.Service{}
+	for _, httpConfig := range httpConfigs {
+		// default group for this ListenOn, in case we don't yet have one yet
+		var groupedConfigs []v2.Service
+		// but replace it with the existing group, if one exists
+		for listenOn, alreadyGrouped := range httpByListenOn {
+			if listenOn == httpConfig.ListenOn {
+				groupedConfigs = alreadyGrouped
+				break
+			}
+		}
+		// append the current config to this ListenOn group
+		groupedConfigs = append(groupedConfigs, httpConfig)
+		httpByListenOn[httpConfig.ListenOn] = groupedConfigs
+	}
+
+	// Now create proper HTTPServiceConfig objects from our map
+	var httpServiceConfigs []HTTPServiceConfig
+	for listenOn, configs := range httpByListenOn {
+		httpServiceConfig := HTTPServiceConfig{
+			SharedListenOn:    listenOn,
+			SubserviceConfigs: configs,
+		}
+		httpServiceConfigs = append(httpServiceConfigs, httpServiceConfig)
+	}
```

Comment:
This is somewhat of a hefty method to parse through in terms of logic - any way to split this up into some smaller functions?

Reply:
Definitely.  I'll plan to do this Monday.